templates/common/_base/files/configure-ovs-network.yaml

```diff
@@ -0,0 +1,116 @@
+      echo "No default route found on attempt: ${counter}"
+      sleep 5
+    done
+
+    if [ "$iface" = "br-ex" ]; then
+      echo "Networking already configured and up for br-ex!"
+      exit 0
+    fi
+
+    if [ -z "$iface" ]; then
+      echo "ERROR: Unable to find default gateway interface"
+      exit 1
+    fi
+
+    # find the MAC from OVS config or the default interface to use for OVS internal port
+    # this prevents us from getting a different DHCP lease and dropping connection
+    if ! iface_mac=$(<"/sys/class/net/${iface}/address"); then
+      echo "Unable to determine default interface MAC"
+      exit 1
+    fi
+
+    echo "MAC address found for iface: ${iface}: ${iface_mac}"
+
+    # find MTU from original iface
+    iface_mtu=$(ip -j link show "$iface" | jq -r '.[0].mtu')
+    if [ -z "$iface_mtu" ]; then
+      echo "Unable to determine default interface MTU, defaulting to 1500"
+      iface_mtu=1500
+    else
+      echo "MTU found for iface: ${iface}: ${iface_mtu}"
+    fi
+
+    # create bridge
+    if ! nmcli connection show br-ex &> /dev/null; then
+      nmcli c add type ovs-bridge conn.interface br-ex con-name br-ex 802-3-ethernet.mtu ${iface_mtu} 802-3-ethernet.cloned-mac-address ${iface_mac}
+    fi
+
+    # store old conn for bringing down later
+    old_conn=$(nmcli --fields UUID,DEVICE conn show --active | grep ${iface} | awk '{print $1}')
+
+    # find default port to add to bridge
+    if ! nmcli connection show ovs-port-phys0 &> /dev/null; then
+      nmcli c add type ovs-port conn.interface ${iface} master br-ex con-name ovs-port-phys0
+    fi
+
+    if ! nmcli connection show ovs-port-br-ex &> /dev/null; then
+      nmcli c add type ovs-port conn.interface br-ex master br-ex con-name ovs-port-br-ex
+    fi
+
+    # bring down any old iface
+    nmcli conn down $old_conn
+
+    if ! nmcli connection show ovs-if-phys0 &> /dev/null; then
+      nmcli c add type 802-3-ethernet conn.interface ${iface} master ovs-port-phys0 con-name ovs-if-phys0 \
+        connection.autoconnect-priority 100 802-3-ethernet.mtu ${iface_mtu}
+    fi
+
+    if ! nmcli connection show ovs-if-br-ex &> /dev/null; then
+      nmcli c add type ovs-interface slave-type ovs-port conn.interface br-ex master ovs-port-br-ex con-name \
+        ovs-if-br-ex 802-3-ethernet.mtu ${iface_mtu} 802-3-ethernet.cloned-mac-address ${iface_mac}
+    fi
+
+    # wait for DHCP to finish, verify connection is up
+    counter=0
+    while [ $counter -lt 5 ]; do
+      sleep 5
+      # check if connection is active
+      if nmcli --fields GENERAL.STATE conn show ovs-if-br-ex | grep -i "activated"; then
+        echo "OVS successfully configured"
+        ip a show br-ex
+        exit 0
+      fi
+      counter=$((counter+1))
+    done
+
+    echo "WARN: OVS did not succesfully activate NM connection. Attempting to bring up connections"
+    counter=0
+    while [ $counter -lt 5 ]; do
+      if nmcli conn up ovs-if-br-ex; then
```

Comment:
Is there some situation where we think the connection won't correctly activate automatically but will correctly activate if we poke at it?

Reply:
Ideally no, but I've hit some inconsistencies in NM + OVS so this is a fail safe to try to address those if we hit them. In general, adding a connection doesn't mean that the connection will automatically connect. For example, if I add a new connection for eth0 that has higher priority than an already activated connection using eth0, it will not bring down the old connection and activate the new.

Doing an nmcli conn up on ovs-if-br-ex (the interface connection) is key because it will bring up all of the other master connections like the OVS port and the bridge and trigger those things to come up as well. I did see a case where doing an nmcli up on ovs-if-br-ex will not bring up ovs-if-phys0 and I had to manually bring that one up. Perhaps in a follow up we can add a check for that as well.